Lthreads disentangled from musl

vtikoo opened the pull request:
### Summary
- Remove musl related threading code from lthreads implementation.
- Add auxiliary vector entries on the elf64 stack created during first ecall.

Fixes #156 

Corresponding sgx-lkl-musl PR: https://github.com/lsds/sgx-lkl-musl/pull/18

vtikoo committed:
point musl submodule to fork

vtikoo committed:
add lthread structures and functions from musl

vtikoo committed:
include lthread header file

vtikoo committed:
move lthread init before libc init

davidchisnall commented on .gitmodules:
```diff
@@ -7,5 +7,5 @@
-	url = https://github.com/lsds/sgx-lkl-musl.git
-	branch = oe_port
+	url = https://github.com/vtikoo/sgx-lkl-musl.git
+	branch = move-lthread-out
```
Don't forget to remove these in the non-draft version!

davidchisnall commented on src/enclave/enclave_event_channel.c:
```diff
@@ -1,12 +1,13 @@
 #include <string.h>
 
-#include "pthread_impl.h"
+// #include "pthread_impl.h"
```
Please just delete these.  We can always find them in the history.

davidchisnall commented on src/enclave/enclave_init.c:
```diff
@@ -240,9 +241,10 @@ int __libc_init_enclave(int argc, char** argv)
-
+    
+    init_ethread_tls();
     __init_libc(envp, argv[0]);
```
Can this be moved after LKL start yet?

davidchisnall commented on src/include/enclave/lthread.h:
```diff
@@ -137,7 +137,7 @@ struct lthread
     struct lthread* lt_join;      /* lthread we want to join on */
     void** lt_exit_ptr;           /* exit ptr for lthread_join */
-    locale_t locale;              /* locale of current lthread */
+    //locale_t locale;              /* locale of current lthread */
```
We should also be able to remove `cancelbuf`, `stdio_locks`, `dlerror_buf`, `dlerror_flag`, `err`.  I'm not sure what `dtv` and `dtv_copy` are, we may also be able to remove them.

davidchisnall commented on src/include/enclave/lthread.h:
```diff
@@ -185,11 +185,83 @@ struct lthread_sched
+	uintptr_t sysinfo;
+	uintptr_t canary, canary2;
+
+	/* Part 2 -- implementation details, non-ABI. */
```
Note that most of these fields are there only because `schedctx` is the same layout as the TLS segment in the old implementation, so that it can live in the same place.  Most of them can be removed.  In fact, I think `lthread_sched` is the only thing that we actually need here (though the initial self pointer is also useful because x86 doesn't give you any other easy way of accessing things in the gs segment).

vtikoo replied:
Do we really need the self pointer? schedctx is located at the end of the ethread gsbase pointed page. And is accessed via a pointer placed at gs:48.

davidchisnall replied:
No, if we're just storing a pointer to this in gs:48 and not the structure in gs, we don't need the self pointer.  The self pointer was there because x86 doesn't have instructions for translating from a segment-relative address to a linear address.  We were storing this structure in the FS segment and the only way of loading a pointer to it was to load the self pointer (%fs:0).  We now have a pointer to it in %gs, so we don't need the self pointer.

davidchisnall commented on src/include/enclave/lthread.h:
```diff
@@ -185,11 +185,83 @@ struct lthread_sched
+    void *self;
+    char _pad_0[32];
+    // SGX-LKL does not have full stack smashing protection (SSP) support right
+    // now. In particular, we do not generate a random stack guard for every
```
Hopefully this comment can be obsolete soon.  Your pthread changes should set one up early and we can generate a canary value for each lthread that isn't created via the primitive call.

davidchisnall commented on src/include/enclave/lthread.h:
```diff
@@ -185,11 +185,83 @@ struct lthread_sched
+    struct schedctx *schedctx;
+};
+
+/* Thread Control Block (TCB) for ethreads/the scheduler (schedctx) */
```
I guess this comment is out of date.  Is this used at all?

vtikoo replied:
This is being used to setup the pointer for schedctx for the ethread in enclave_oe.c. But that can probably be moved to use lthread_tcb_base or some inline assembly. Any thoughts?

davidchisnall commented on src/include/enclave/lthread.h:
```diff
@@ -185,11 +185,83 @@ struct lthread_sched
 extern "C"
 {
 #endif
+    void init_ethread_tls();
```
Please add a doc comment here.

davidchisnall commented on src/include/enclave/lthread_int.h:
```diff
@@ -71,4 +71,20 @@ static inline uint64_t _lthread_timespec_to_usec(const struct timespec* ts)
 }
 
+// TODO: should this be static?
+static inline struct schedctx *__scheduler_self()
```
Please add a doc comment.  It should be `static inline` so that it can be inlined everywhere (it's a single instruction after inlining) and the compiler won't complain about it being in a header.

vtikoo replied:
Added docs.

davidchisnall commented on src/include/enclave/lthread_int.h:
```diff
@@ -71,4 +71,20 @@ static inline uint64_t _lthread_timespec_to_usec(const struct timespec* ts)
+static inline struct lthread_sched*
+lthread_get_sched()
+{
+    struct schedctx *c = __scheduler_self();
```
Are there noe any other callers of `__scheduler_self`?  If not, it can be inlined here.

vtikoo replied:
Its also used to setup the `schedctx` struct in `enclave_oe.c:init_ethread_tp`.

davidchisnall commented on src/sched/lthread.c:
```diff
@@ -63,7 +63,33 @@ static void lthread_rundestructors(struct lthread* lt);
+
+	/* Failure to initialize thread pointer is always fatal. */
+	if (__init_ethread_tp(mem) < 0)
+		a_crash();
```
This is dead code, `__init_ethread_tp` unconditionally returns 0 so this is never reached.

It's probably a good idea to simply inline `__init_ethread_tp` here.

davidchisnall commented on src/sched/lthread.c:
```diff
@@ -461,6 +487,50 @@ void _lthread_free(struct lthread* lt)
+	{
+		if (sgxlkl_enclave->mode == SW_DEBUG_MODE)
+		{
+			int r = __set_thread_area(TP_ADJ(p));
```
As I recall, this is implemented in musl.  If so, we need to remove it.  I believe that, in software mode, this should do a host syscall to set the %fs register value.

davidchisnall commented on src/sched/lthread.c:
```diff
@@ -461,6 +487,50 @@ void _lthread_free(struct lthread* lt)
+	struct lthread_tcb_base *tcb = (struct lthread_tcb_base *)p;
+	tcb->self = p;
+	tcb->schedctx = __scheduler_self();
+	if (libc.user_tls_enabled && set_tp)
```
The first of these checks should be removed.  We want to set up a TLS area with just a stack canary for threads that are not created with `lthread_create_primitive`.

vtikoo replied:
`lthread_create` calls this function with set_tp set to 0. Should we remove the entire block?

davidchisnall commented on src/sched/lthread.c:
```diff
@@ -461,6 +487,50 @@ void _lthread_free(struct lthread* lt)
+	return 0;
+}
+
+void *__copy_utls(struct lthread *lt, unsigned char *mem, size_t sz)
```
Most of this should go away.  Threads created in musl will handle this, so this is just for kernel and host-interface threads.  The only thing that they need is a new random stack canary value.

vtikoo replied:
Removing both copy_utls and __init_utp, as they are effectively serving no purpose currently. The only usage in lthread_create calls __init_utp with set_tp = 0, at which point both of these functions are only doing pointer manipulations on local variables.

davidchisnall commented on src/sched/lthread.c:
```diff
@@ -637,47 +707,37 @@ int lthread_create_primitive(
+    //Rm tls_size from lt 
+    // lt->itlssz = libc.tls_size;
+    lt->itls = tls;
+    // if (libc.tls_size)
```
You can also remove most of the equivalent changes in `lthread_create`.  That function should only be called for non-userspace threads and so doesn't need to touch any libc state.

davidchisnall reviewed (commented):
Thanks for working on this, looks like a great start.

vtikoo committed:
userspace thread should be based of fs

jxyang commented on src/sched/lthread.c:
```diff
@@ -732,16 +792,16 @@ int lthread_create(
-        init_file_lock(__stderr_used);
-        libc.threaded = 1;
-    }
+    // if (!libc.threaded && libc.threads_minus_1 >= 0)
```
We don't need to initialize the locks because `pthread_create` does this? What about the lthreads we created directly without going through libc?

davidchisnall replied:
We will be doing that only for kernel threads, which should not hold any userspace libc locks.

letmaik commented:
@vtikoo Could you set the PR to draft mode if it's not ready for merging yet?

vtikoo committed:
add docs

vtikoo committed:
cleanup locale, robust_list, stdio_locks changes

vtikoo committed:
remove dtv related code

vtikoo committed:
set tls size for userspace threads

vtikoo committed:
- set tls area for host threads
- skip tp calculation for userspace threads

vtikoo committed:
add tls and conditional variable test

davidchisnall commented on src/sched/lthread.c:
```diff
@@ -474,23 +476,32 @@ int __init_utp(void *p, int set_tp)
-    uintptr_t tp_unaligned =
+    // pthread_create passes thread pointer for lt->itls
+    // no need to calculate offset of tp in tls region
+    if (lt->attr.stack_size == 0)
```
Sorry, I may have just broken this logic with the deferred stack freeing PR.  It's probably worth adding a flag in the lthread structure to check if this is a cloned task or not.  That said, I'd prefer if we always just set `tp = lt->itls` unconditionally and modified the other code path that set itls to do the adjustment.

vtikoo replied:
Adding thread_type to lthread_attr

davidchisnall commented on src/sched/lthread.c:
```diff
@@ -474,23 +476,32 @@ int __init_utp(void *p, int set_tp)
+            (struct
+            lthread_tcb_base*)(tp_unaligned - (tp_unaligned & (TLS_ALIGN - 1)));
 
+        tp->schedctx = __scheduler_self();
```
This should no longer be required.  I don't think we ever read this field, we should always be getting the scheduler via %gs.  If you remove this, does anything break?

vtikoo replied:
Should be safe to remove, will do so.

davidchisnall commented on src/sched/lthread.c:
```diff
@@ -664,6 +675,7 @@ int lthread_create_primitive(
     lt->itlssz = libc.tls_size; // not setting this causes problems in set_tls_tp
```
I think your changes above make this unnecessary.

davidchisnall commented on src/sched/lthread.c:
```diff
@@ -664,6 +675,7 @@ int lthread_create_primitive(
     lt->itlssz = libc.tls_size; // not setting this causes problems in set_tls_tp
     LIST_INIT(&lt->tls);
     lt->attr.state = BIT(LT_ST_READY);
+    lt->attr.stack_size = 0; // lthread doesn't manage userspace thread stacks
```
The lthread should be allocated with `oe_calloc`, so this should be dead code.

davidchisnall commented on src/sched/lthread.c:
```diff
@@ -733,8 +745,8 @@ int lthread_create(
     /* mmap tls image */
-    lt->itlssz = libc.tls_size;
-    if (libc.tls_size)
+    lt->itlssz = (sizeof(lthread_tcb_base) + TLS_ALIGN - 1) & -TLS_ALIGN;
+    if (lt->itlssz)
```
This logic looks confusing now.  We're setting `lt->itlssz` to a constant and then branching on that constant being non-zero.  I think the body of this if statement is reached unconditionally, so we should just remove this line.

vtikoo replied:
Good point!

davidchisnall commented on src/sched/lthread.c:
```diff
@@ -733,8 +745,8 @@ int lthread_create(
+    lt->itlssz = (sizeof(lthread_tcb_base) + TLS_ALIGN - 1) & -TLS_ALIGN;
+    if (lt->itlssz)
     {
         if ((lt->itls = (uint8_t*)enclave_mmap(
```
If we do the adjustment once here (which is a very cold code path), we can avoid the conditional in the `set_tls_tp` function (which is a hot path, hit on every context switch).

vtikoo committed:
cleanup lthread tls handling

davidchisnall commented on src/sched/lthread.c:
```diff
@@ -672,10 +605,10 @@ int lthread_create_primitive(
 
     lt->itls = tls;
-    lt->itlssz = libc.tls_size; // not setting this causes problems in set_tls_tp
+    lt->tp = tls; // for cloned threads, the tls argument is a ptr to the TCB
```
Please expand TCB: in a project like this, Thread Control Block and Trusted Computing Base are both in use and the initialism is confusing!

vtikoo committed:
cleanup __init_tp for kernel threads

vtikoo committed:
add docs for __scheduler_self

vtikoo committed:
remove sched_tcb_base references

vtikoo committed:
remove unused schedctx fields

vtikoo committed:
remove unused lthread fields

vtikoo committed:
remove pthread header files

vtikoo committed:
remove unused var

davidchisnall commented on src/sched/lthread.c:
```diff
@@ -580,13 +588,13 @@ int _lthread_sched_init(size_t stack_size)
+    sched->default_timeout = 3000000u;
 
-    memset(&c->sched.ctx, 0, sizeof(struct cpu_ctx));
+    memset(&sched->ctx, 0, sizeof(struct cpu_ctx));
```
Not very important, but I wonder if this is actually necessary.  This field should be fully set by `_switch` the first time this scheduler runs an lthread and the zero value should never be read.

davidchisnall commented on src/sched/lthread.c:
```diff
@@ -605,14 +613,16 @@ int lthread_create_primitive(
 
     static unsigned long long n = 0;
-    snprintf(lt->funcname, 64, "cloned host task %llu", __atomic_fetch_add(&n, 1, __ATOMIC_SEQ_CST));
+    snprintf(lt->funcname, 64, "cloned host task %llu",
```
This probably should be `oe_snprintf` (my fault, I think).

davidchisnall reviewed (commented):
Diffs that mostly delete code make me happy!

vtikoo committed:
add test for sched_yield

vtikoo committed:
set tls self pointer for cloned threads

vtikoo committed:
remove strerror usage

vtikoo committed:
set locale in startmain thread for userspace components

davidchisnall commented on src/enclave/enclave_init.c:
```diff
@@ -148,6 +148,9 @@ static int startmain(void* args)
 
+    /* Set locale for usersapce components using it */
+    pthread_t self = __pthread_self();
+    self->locale = &libc.global_locale;
```
This setup should be done in the start libc routine in musl.

vtikoo replied:
Oh okay, makes sense. I'll look for the appropriate place within musl to set this.

jxyang replied:
This is for the `startmain` lthread which is created before libc setup, right? Are you trying to wrap it inside a pthread and patch it up? Where and why do we expect `pthread->locale` to be set?

davidchisnall commented on tests/basic/clone/clone.c:
```diff
@@ -125,7 +125,7 @@ int main(int argc, char** argv)
 	assert(ctid == clone_ret, "ctid is %d, should be %d\n", ctid, clone_ret);
 	int futex_ret = futex_wait(&ctid, clone_ret);
-	assert(futex_ret == 0, "futex syscall returned %d (%s)\n", strerror(errno));
+	assert(futex_ret == 0, "futex syscall returned %d\n", futex_ret);
```
The format string was wrong (ooops, my fault, sorry!), but this is in the main thread so strerror should be safe to use.

davidchisnall commented on src/sched/lthread.c:
```diff
@@ -566,6 +566,9 @@ int lthread_create_primitive(
     lt->tp = tls;
+    // set self pointer for fs segment
+    struct lthread_tcb_base *tcb = (struct lthread_tcb_base *)lt->tp;
+    tcb->self = lt->tp;
```
Musl should be doing this for pthreads, is this just needed for the clone test? If so, we should probably fix that test (I think it’s actually fine to remove it and replace it with a pthread test that does the same thing, it’s only needed because pthreads weren’t using clone when I wrote it and we needed to test clone.).

vtikoo replied:
`tests/basic/pthread_tls` probably covers the behaviors the clone test covers - userspace thread creation and waiting on futex. It can't cover the deferred stack unmapping test though.

jxyang replied:
The self pointer is already set by musl here: https://github.com/lsds/sgx-lkl-musl/blob/f437aa45469273c414ceba14c884ec7d126f912a/src/thread/pthread_create.c#L265

vtikoo replied:
This was added to support the clone test, I'll move this logic there. Should we add an assert here to check if the self pointer has been set?

davidchisnall commented:
Please can you create a draft musl PR for the musl changes so that I can review them as well?  Thanks.

vtikoo commented:
@davidchisnall the musl PR is here - https://github.com/lsds/sgx-lkl-musl/pull/18

vtikoo committed:
Save fs pointer for initial lthread

vtikoo committed:
Add debug targets

vtikoo committed:
Fix rebasing issues

vtikoo committed:
Don't free config in the termination thread

vtikoo committed:
Move writing fsbase into separate function

vtikoo committed:
Don't cleanup detached userspace threads.

LKL handles clean for threads created via clone() during its termination sequence.

vtikoo committed:
Disable rt_sigqueueinfo01 test

vtikoo committed:
Move enclave config cleanup into separate function

vtikoo committed:
Setup aux vector during _prepare_elf_stack

vtikoo committed:
Initialize heap early

vtikoo committed:
Call init_libc before loading the app

vtikoo committed:
Move libc init before LKL kernel init

vtikoo committed:
Disable setreuid07 and setresuid04 until corresponding ltp patch is merged

vtikoo commented on .azure-pipelines/other/disabled_tests.txt:
```diff
@@ -1,3 +1,4 @@
+tests/basic/pthread_join/Makefile
```
This test is disabled until https://github.com/lsds/sgx-lkl/issues/686 is fixed.

vtikoo commented on src/enclave/enclave_init.c:
```diff
@@ -90,6 +91,8 @@ static void init_wireguard()
 
 static int startmain(void* args)
 {
+    __init_libc(sgxlkl_enclave_state.elf64_stack.envp,
```
This can't be moved any further than this right now. `lkl_start_init` does some `getenv` calls which requires libc to be initialized.

mikbras replied:
Don't worry. My feature.split branch will re-organize this after check in.

jxyang commented on src/enclave/enclave_event_channel.c:
```diff
@@ -1,12 +1,11 @@
 
 #include "enclave/enclave_util.h"
 #include "enclave/lthread.h"
+#include "enclave/lthread_int.h"
```
Why do we need the internal lthread header file?

vtikoo replied:
Its relying on it for two macros: `BIT` and `CLEARBIT`, and `_lthread_yield_cb()`.

jxyang commented on src/enclave/enclave_init.c:
```diff
@@ -150,9 +156,10 @@ int __libc_init_enclave(int argc, char** argv)
 
-    __init_libc(envp, argv[0]);
-    __init_tls();
+    __init_heap_from_libc();
```
Maybe some explanation on why this is needed.

vtikoo replied:
Sure. I had added some documentation in `__libc_startmain.c` in musl, where this function is defined. Will add here as well. This should anyway be a temporary fix until memory corruption issues are root caused.

jxyang commented on src/enclave/enclave_event_channel.c:
```diff
@@ -1,12 +1,11 @@
 
 #include "enclave/enclave_util.h"
 #include "enclave/lthread.h"
+#include "enclave/lthread_int.h"
```
Why do we need this internal header file?

jxyang commented on src/enclave/enclave_oe.c:
```diff
@@ -399,3 +452,12 @@ void sgxlkl_free_enclave_state()
     _free_shared_memory();
 }
+
+void cleanup_sgxlkl_enclave_config()
```
Seems this function is not called anywhere?

vtikoo replied:
Resolved with - https://github.com/vtikoo/sgx-lkl/commit/dd89071da31cdeae5217784c2f4cb5eaecd3bbf3

jxyang commented on src/sched/lthread.c:
```diff
@@ -294,10 +308,12 @@ void _lthread_desched_sleep(struct lthread* lt)
     ticket_unlock(&sleeplock);
 
     SGXLKL_TRACE_THREAD(
```
Why do we need two log messages for this simple function?

jxyang commented on src/sched/lthread.c:
```diff
@@ -337,45 +353,27 @@ void _lthread_yield(struct lthread* lt)
         lthread_rundestructors(lt);
-    if (lt->itls != 0)
+    
+    // lthread only manages tls region for lkl kernel threads
```
Do we really have kernel lthread using TLS?

vtikoo replied:
Thats a good question, I'm not sure. @davidchisnall do you know if we need `itls` for kernel threads?

davidchisnall replied:
We need the stack canary value, that should be all.  Eventually, we should just populate a page or two with random numbers on boot and pick a random offset in that region for the itls pointer such that the stack canary lines up somewhere inside it, but for now we need one.

vtikoo replied:
We also need a self pointer in `itls` for `set_tls_tp` to work.

jxyang commented on src/sched/lthread.c:
```diff
@@ -590,51 +590,21 @@ int lthread_create_primitive(
-            return -1;
-        }
-    }
+    lt->itls = tls;
```
For my understanding, why do we need two pointers to the same TLS?

jxyang commented on src/sched/lthread.c:
```diff
@@ -720,34 +682,26 @@ int lthread_create(
-    if (libc.tls_size)
-    {
-        if ((intptr_t)(lt->itls = (uint8_t*)enclave_mmap(
+    lt->itlssz = (sizeof(struct lthread_tcb_base) + TLS_ALIGN - 1) & -TLS_ALIGN;
```
Can you explain how do we calculate the itlssz here, probably with a comment?

vtikoo replied:
Added comment.

jxyang commented on src/sched/lthread.c:
```diff
@@ -494,10 +493,23 @@ int _lthread_resume(struct lthread* lt)
+        void* fs_ptr;
+        __asm__ __volatile__("mov %%fs:0,%0" : "=r"(fs_ptr));
+        if (fs_ptr != sched->current_lthread->tp){
+            sched->current_lthread->tp = fs_ptr;
```
Isn't this initialized during lthread_create (with the init_tp call)?

vtikoo replied:
`init_tp` calculates the thread pointer value and sets a self pointer, but it does not set fsbase. FS register is set during context switches in `set_tls_tp`.

This above code is required because for the `startmain` thread, the dynamic loader creates a new tls region before passing execution to it - [here](https://github.com/vtikoo/sgx-lkl-musl/blob/move-lthread-out/ldso/dynlink.c#L1951-L1959).
lthreads needs to save that new pointer.

jxyang commented:
I will take a look of the musl changes later.

vtikoo committed:
Rebase with oe_port

vtikoo committed:
Add comment for temporary workaround

vtikoo committed:
Address Issue# 742 by requesting a page early

vtikoo committed:
Free enclave config when last ethread exits.

lthread scheduler uses the sgxlkl_enclave_state.config->mode for context switching.
So, only free once we are sure all schedulers have shut down.

vtikoo committed:
Set self pointer in tls region

vtikoo committed:
Revert change to context switch epilogue

vtikoo committed:
Add test creating 100 detached threads sequentially

vtikoo committed:
Update sgx-lkl-musl submodule

vtikoo committed:
Fix type of thread pointer

vtikoo committed:
Update ltp submodule commit

vtikoo committed:
Enable setresuid04 and setreuid07

vtikoo committed:
Update ltp submodule commit

vtikoo committed:
Disable setresuid04 and setreuid07

vtikoo committed:
Revert sgx-lkl-musl remote to lsds:oe_port

vtikoo committed:
Revert enclave config late cleanup changes

vtikoo committed:
Remove fix for #742

vtikoo committed:
Address PR comment

vtikoo committed:
Set self pointer in tls region

vtikoo committed:
Simplify tp setting logic for cloned threads

vtikoo committed:
Update sgx-lkl-musl submodule

vtikoo committed:
Add nodejs and pthread_detach tests to disabled

mikbras commented on src/enclave/enclave_init.c:
```diff
@@ -150,9 +156,8 @@ int __libc_init_enclave(int argc, char** argv)
     newmpmcq(&__scheduler_queue, max_lthreads, 0);
-
-    __init_libc(envp, argv[0]);
-    __init_tls();
```
Just double checking. Has the call to this function gone away completely? (__init_tls)?

mikbras commented on src/enclave/enclave_oe.c:
```diff
@@ -48,6 +55,54 @@ static int _strncmp(const char* x, const char* y, size_t n)
+{
+    // By default auxv[AT_RANDOM] points to a buffer with 16 random bytes.
+    uint64_t *rbuf = (uint64_t*)buf_ptr;
+    buf_ptr += 16;
```
Is this buffer 16-byte aligned? I think the ABI requires it.

mikbras commented on src/enclave/enclave_oe.c:
```diff
@@ -48,6 +55,54 @@ static int _strncmp(const char* x, const char* y, size_t n)
+    auxv[23] = !sgxlkl_in_sw_debug_mode();
+    auxv[24] = AT_NULL;
+    auxv[25] = 0;
+}
```
The following seem to be missing: 
- AT_BASE -- base address of image
- AT_SYSINFO_EHDR -- the ELF header address
- AT_PHDR - the program header address
- AT_PHNUM - the number of program headers
 - AT_ENTRY - address of the entry point

These are usually passed by the loader. It may work without them for now, but these may be needed later when we want to just call _start().

You can check in without these I think.

davidchisnall replied:
`AT_PHDR` is needed for unwind libraries.  We probably can't throw an exception through `qsort` without that.  That said, I've never actually seen code in the wild that wants to throw exceptions through libc functions outside of test suites: if you're using C++, there are significantly more efficient ways of sorting than calling libc's `qsort` and very few other libc functions take function pointer arguments.

mikbras reviewed (changes requested):
Vikas. If you can verify that auxv is 16-byte aligned, I will approve. Also I notice that some of the standard AT_* variables are missing in auxv. This is fine for now but when we back out certain musl libc changes, they will be needed to find the elf header, program headers, and the elf entry point.

vtikoo commented:
As discussed with @mikbras, #325 covers the effort to get to full ABI compatibility for the Stack on app entry.

vtikoo committed:
Update sgx-lkl-musl submodule

vtikoo committed:
Re-enable setresuid04 and setreuid07
